In [2]:
conda install -c anaconda beautifulsoup4

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Gary\Anaconda3

  added / updated specs:
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    beautifulsoup4-4.7.1       |           py37_1         143 KB  anaconda
    ca-certificates-2019.5.15  |                0         166 KB  anaconda
    certifi-2019.6.16          |           py37_0         155 KB  anaconda
    conda-4.7.12               |           py37_0         3.0 MB  anaconda
    openssl-1.1.1c             |       he774522_1         5.7 MB  anaconda
    ------------------------------------------------------------
                                           Total:         9.2 MB

The following packages will be UPDATED:

  conda                      pkgs/main::conda-4.7.10-py37_0 --> anaconda::conda-4.7.12-py37_0

The following packages will be SUPERSEDED by a higher-pr



==> WARNING: A newer version of conda exists. <==
  current version: 4.7.10
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda




In [2]:
conda update -n base -c defaults conda


Note: you may need to restart the kernel to use updated packages.


usage: conda-script.py install [-h] [--revision REVISION]
                               [-n ENVIRONMENT | -p PATH] [-c CHANNEL]
                               [--use-local] [--override-channels]
                               [--repodata-fn REPODATA_FNS]
                               [--strict-channel-priority]
                               [--no-channel-priority]
                               [--no-deps | --only-deps] [--no-pin] [--copy]
                               [--no-shortcuts] [-C] [-k] [--offline] [-d]
                               [--json] [-q] [-v] [-y] [--download-only]
                               [--show-channel-urls] [--file FILE]
                               [--force-reinstall]
                               [--freeze-installed | --update-deps | -S | --update-all | --update-specs]
                               [-m] [--clobber] [--dev]
                               [package_spec [package_spec ...]]
conda-script.py install: error: argument -n/--name: not allow

In [3]:
conda install -c anaconda html5lib

Solving environment: ...working... 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/win-64::anaconda==2019.07=py37_0
  - defaults/win-64::numba==0.44.1=py37hf9181ef_0
done

## Package Plan ##

  environment location: C:\Users\Gary\Anaconda3

  added / updated specs:
    - html5lib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _anaconda_depends-2019.03  |           py37_0           5 KB  anaconda
    anaconda-custom            |           py37_1           3 KB
    ca-certificates-2019.8.28  |                0         165 KB  anaconda
    certifi-2019.9.11          |           py37_0         155 KB  anaconda
    html5lib-1.0.1             |           py37_0         181 KB  anaconda
    openssl-1.1.1              |       he774522_0         5.7 MB  anaconda
    tbb-2019.4                 |       

In [4]:
conda install -c anaconda requests

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Gary\Anaconda3

  added / updated specs:
    - requests


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    requests-2.22.0            |           py37_0          90 KB  anaconda
    ------------------------------------------------------------
                                           Total:          90 KB

The following packages will be SUPERSEDED by a higher-priority channel:

  requests                                        pkgs/main --> anaconda




requests-2.22.0      | 90 KB     |            |   0% 
requests-2.22.0      | 90 KB     | #7         |  18% 
requests-2.22.0      | 90 KB     | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done

Note: you may need to restart the kernel to use updated pac

In [2]:
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Gary\Anaconda3

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    altair-3.2.0               |           py37_0         749 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         181 KB  conda-forge
    certifi-2019.6.16          |           py37_1         149 KB  conda-forge
    conda-4.7.12               |           py37_0         3.0 MB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1c             |       hfa6e2cd_0         4.7 MB  conda-forge
    vincent-0.4.4              |             py_1      

# Scrape the Wikipedia page and read the table of postal codes and to transform the data into a pandas dataframe 

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source=requests.get(url).text
soup=BeautifulSoup(source,'lxml')

In [3]:
table_page=soup.find('div',class_='mw-parser-output')

In [4]:
header_items=table_page.table.tr.text
header_items=header_items.split('\n')[1:4]

In [5]:
table_value=table_page.table.tbody
table_value=table_value.find_all('td')

In [6]:
Pos=[]
Bor=[]
Nei=[]
for n in range(0,len(table_value),3):
    Pos.append(table_value[n].text)
    Bor.append(table_value[n+1].text)
    Nei.append(table_value[n+2].text)

In [7]:
for n in range(len(Nei)):
    Nei[n]=Nei[n][:-1]

In [8]:
for i in range(len(Nei)):
    if Nei[i]=='Not assigned':
        Nei[i]=Bor[i]

In [9]:
final_array=[]
for pos, bor,nei in zip(Pos,Bor,Nei):
    final_array.append({header_items[0]:pos,header_items[1]:bor,header_items[2]:nei})

In [10]:
import pandas as pd
df= pd.DataFrame(final_array)

In [11]:
df.rename(columns={'Postcode':'PostalCode','Neighbourhood':'Neighborhood'},inplace=True)

In [12]:
df=df[['PostalCode','Borough','Neighborhood']]
df=df[df['Borough']!='Not assigned']

In [13]:
def combine(df):
    return ','.join(df.values)
df2=df.groupby(['PostalCode','Borough'])['Neighborhood'].apply(combine)
df2=df2.to_frame()
df2.reset_index(inplace=True)
df2

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                        Rouge,Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7                        Clairlea,Golden Mile,Oakridge  
8        Cliffcrest,Cliffside,Scarborough Village West  
9                           Birch Cliff,Cliffside West  
10   Dorset Park,Scarborough Town Centre,Wexford He...  
11                                    Maryvale,Wexford  
12                                           Agincourt  
13               Clarks Corners,Sullivan,Tam O'Shanter  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                          Fairview,Henry Farm,Oriole  
19                                     Bayview Village  
20                             Silver Hills,York Mills  
21                              Newtonbrook,Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                     Flemingdon Park,Don Mills South  
28       Bathurst Manor,Downsview North,Wilson Heights  
29                      Northwood Park,York University  
..  

In [14]:
df2.shape

(103, 3)

The table of postal codes contains 103 distinct rows.

# Use the Geocoder csv file to create a new dataframe with latitude and longitude:

In [15]:
df_cd=pd.read_csv('https://cocl.us/Geospatial_data')
df_cd.rename(columns={'Postal Code':'PostalCode'},inplace=True)

In [16]:
df_final=pd.merge(df_cd,df2,on='PostalCode')
df_final=df_final[['PostalCode','Borough','Neighborhood','Latitude','Longitude']]
df_final

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6            East Birchmount Park,Ionview,Kennedy Park  43.727929 -79.262029  
7                        Clairlea,Golden Mile,Oakridge  43.711112 -79.284577  
8        Cliffcrest,Cliffside,Scarborough Village West  43.716316 -79.239476  
9                           Birch Cliff,Cliffside West  43.692657 -79.264848  
10   Dorset Park,Scarborough Town Centre,Wexford He...  43.757410 -79.273304  
11                                    Maryvale,Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13               Clarks Corners,Sullivan,Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                          Fairview,Henry Farm,Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                             Silver Hills,York Mills  43.757490 -79.374714  
21                              N

In [17]:
df_final.shape

(103, 5)

In [18]:
df_map=df_final[df_final['Borough'].str.contains('Toronto')]
df_map

PostalCode           Borough  \
37        M4E      East Toronto   
41        M4K      East Toronto   
42        M4L      East Toronto   
43        M4M      East Toronto   
44        M4N   Central Toronto   
45        M4P   Central Toronto   
46        M4R   Central Toronto   
47        M4S   Central Toronto   
48        M4T   Central Toronto   
49        M4V   Central Toronto   
50        M4W  Downtown Toronto   
51        M4X  Downtown Toronto   
52        M4Y  Downtown Toronto   
53        M5A  Downtown Toronto   
54        M5B  Downtown Toronto   
55        M5C  Downtown Toronto   
56        M5E  Downtown Toronto   
57        M5G  Downtown Toronto   
58        M5H  Downtown Toronto   
59        M5J  Downtown Toronto   
60        M5K  Downtown Toronto   
61        M5L  Downtown Toronto   
63        M5N   Central Toronto   
64        M5P   Central Toronto   
65        M5R   Central Toronto   
66        M5S  Downtown Toronto   
67        M5T  Downtown Toronto   
68        M5V  Downtown Toronto   
69        M5W  Downtown Toronto   
70        M5X  Downtown Toronto   
75        M6G  Downtown Toronto   
76        M6H      West Toronto   
77        M6J      West Toronto   
78        M6K      West Toronto   
82        M6P      West Toronto   
83        M6R      West Toronto   
84        M6S      West Toronto   
87        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                        The Danforth West,Riverdale  43.679557 -79.352188  
42                      The Beaches West,India Bazaar  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                                 North Toronto West  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                         Moore Park,Summerhill East  43.689574 -79.383160  
49  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                         Cabbagetown,St. James Town  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                           Harbourfront,Regent Park  43.654260 -79.360636  
54                            Ryerson,Garden District  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                             Adelaide,King,Richmond  43.650571 -79.384568  
59    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
60            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
61                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
65                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
66                      Harbord,University of Toronto  43.662696 -79.400049  
67            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
68  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
69                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
70              First Canadian Place,Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                        Dovercourt Village,Dufferin  43.669005 -79.442259  
77                            Little Portugal,Trinity  43.64792

In [19]:
CLIENT_ID = 'QPG4LS2XMXGC2KXOGFMHKXCTIACEGEYEKGT0BS3E0DAOXGDG' # your Foursquare ID
CLIENT_SECRET = 'WVWNOOZAJ11DOKZAKJWKPPD1LRRESFJFZ4BABL5RBDD50UT1' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QPG4LS2XMXGC2KXOGFMHKXCTIACEGEYEKGT0BS3E0DAOXGDG
CLIENT_SECRET:WVWNOOZAJ11DOKZAKJWKPPD1LRRESFJFZ4BABL5RBDD50UT1


In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    LIMIT = 100
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
map_result=getNearbyVenues(names=df_map['Neighborhood'],
                           latitudes=df_map['Latitude'],
                          longitudes=df_map['Longitude'])

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvall

In [22]:
map_result

Neighborhood  \
0                                           The Beaches   
1                                           The Beaches   
2                                           The Beaches   
3                                           The Beaches   
4                                           The Beaches   
5                           The Danforth West,Riverdale   
6                           The Danforth West,Riverdale   
7                           The Danforth West,Riverdale   
8                           The Danforth West,Riverdale   
9                           The Danforth West,Riverdale   
10                          The Danforth West,Riverdale   
11                          The Danforth West,Riverdale   
12                          The Danforth West,Riverdale   
13                          The Danforth West,Riverdale   
14                          The Danforth West,Riverdale   
15                          The Danforth West,Riverdale   
16                          The Danforth West,Riverdale   
17                          The Danforth West,Riverdale   
18                          The Danforth West,Riverdale   
19                          The Danforth West,Riverdale   
20                          The Danforth West,Riverdale   
21                          The Danforth West,Riverdale   
22                          The Danforth West,Riverdale   
23                          The Danforth West,Riverdale   
24                          The Danforth West,Riverdale   
25                          The Danforth West,Riverdale   
26                          The Danforth West,Riverdale   
27                          The Danforth West,Riverdale   
28                          The Danforth West,Riverdale   
29                          The Danforth West,Riverdale   
...                                                 ...   
1685                                  Runnymede,Swansea   
1686                                  Runnymede,Swansea   
1687                                  Runnymede,Swansea   
1688                                  Runnymede,Swansea   
1689                                  Runnymede,Swansea   
1690                                  Runnymede,Swansea   
1691                                  Runnymede,Swansea   
1692                                  Runnymede,Swansea   
1693                                  Runnymede,Swansea   
1694                                  Runnymede,Swansea   
1695                                  Runnymede,Swansea   
1696  Business Reply Mail Processing Centre 969 Eastern   
1697  Business Reply Mail Processing Centre 969 Eastern   
1698  Business Reply Mail Processing Centre 969 Eastern   
1699  Business Reply Mail Processing Centre 969 Eastern   
1700  Business Reply Mail Processing Centre 969 Eastern   
1701  Business Reply Mail Processing Centre 969 Eastern   
1702  Business Reply Mail Processing Centre 969 Eastern   
1703  Business Reply Mail Processing Centre 969 Eastern   
1704  Business Reply Mail Processing Centre 969 Eastern   
1705  Business Reply Mail Processing Centre 969 Eastern   
1706  Business Reply Mail Processing Centre 969 Eastern   
1707  Business Reply Mail Processing Centre 969 Eastern   
1708  Business Reply Mail Processing Centre 969 Eastern   
1709  Business Reply Mail Processing Centre 969 Eastern   
1710  Business Reply Mail Processing Centre 969 Eastern   
1711  Business Reply Mail Processing Centre 969 Eastern   
1712  Business Reply Mail Processing Centre 969 Eastern   
1713  Business Reply Mail Processing Centre 969 Eastern   
1714  Business Reply Mail Processing Centre 969 Eastern   

      Neighborhood Latitude  Neighborhood Longitude  \
0                 43.676357              -79.293031   
1                 43.676357              -79.293031   
2                 43.676357              -79.293031   
3                 43.676357              -79.293031   
4                 43.676357              -79.293031   
5                 43.679557              -79.352188   

In [23]:
map_result.shape

(1715, 7)

In [24]:
# venues summary
toronto_venues = pd.get_dummies(map_result[['Venue Category']], prefix="", prefix_sep="")

In [25]:
#add Neighbor into the dataframe to avoid overlapping.
## there is a column called Neighborhood
toronto_venues.insert(0,'Neighbor',map_result['Neighborhood'])

In [26]:
toronto_grouped = toronto_venues.groupby('Neighbor').mean().reset_index()
toronto_grouped

Neighbor  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011494   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19    Harbourfront East,Toronto Islands,Union Station           0.000000   
20                           Harbourfront,Regent Park           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                           Rosedale           0.000000   
28                                           Roselawn           0.000000   
29                                  Runnymede,Swansea           0.000000   
30                            Ryerson,Garden District           0.000000   
31                                     St. James Town           0.000000   
32                    Stn A PO Boxes 25 The Esplanade           0.000000   
33                                    Studio District           0.000000   
34                  The Annex,North Midtown,Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                      The Beaches West,India Bazaar           0.000000   
37                        The Danforth West,Riverdale           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.066667            0.066667      0.066667        0.133333   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000            

# Use k-means to cluster neighborhood of Toronto by venues characteristics.

In [27]:
from sklearn.cluster import KMeans

In [28]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbor', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[-10:-1] 

array([4, 3, 3, 3, 3, 3, 3, 3, 3])

In [29]:
# add clustering labels
toronto_grouped.insert(1, 'Cluster Labels', kmeans.labels_)

In [30]:
toronto_merged = df_map
toronto_merged.rename(columns={'Neighborhood':'Neighbor'},inplace=True)
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_grouped.set_index('Neighbor'), on='Neighbor')

toronto_merged.head() # check the last columns!

C:\Users\Gary\Anaconda3\lib\site-packages\pandas\core\frame.py:4025: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


PostalCode          Borough                       Neighbor   Latitude  \
37        M4E     East Toronto                    The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
42        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
43        M4M     East Toronto                Studio District  43.659526   
44        M4N  Central Toronto                  Lawrence Park  43.728020   

    Longitude  Cluster Labels  Afghan Restaurant  Airport  Airport Food Court  \
37 -79.293031               3                0.0      0.0                 0.0   
41 -79.352188               3                0.0      0.0                 0.0   
42 -79.315572               3                0.0      0.0                 0.0   
43 -79.340923               3                0.0      0.0                 0.0   
44 -79.388790               2                0.0      0.0                 0.0   

    Airport Gate  ...  Toy / Game Store     Trail  Train Station  \
37           0.0  ...               0.0  0.200000            0.0   
41           0.0  ...               0.0  0.023256            0.0   
42           0.0  ...               0.0  0.000000            0.0   
43           0.0  ...               0.0  0.000000            0.0   
44           0.0  ...               0.0  0.000000            0.0   

    Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
37                            0.0               0.0                    0.0   
41                            0.0               0.0                    0.0   
42                            0.0               0.0                    0.0   
43                            0.0               0.0                    0.0   
44                            0.0               0.0                    0.0   

    Wine Bar  Wings Joint  Women's Store  Yoga Studio  
37       0.0          0.0            0.0     0.000000  
41       0.0          0.0            0.0     0.023256  
42       0.0          0.0            0.0     0.000000  
43       0.0          0.0            0.0     0.025641  
44       0.0          0.0            0.0     0.000000  

[5 rows x 246 columns]

In [31]:
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors

In [33]:
import folium
# create map
map_final = folium.Map(location=[43.6532, -79.3832], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbor'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-2],
        fill=True,
        fill_color=rainbow[cluster-2],
        fill_opacity=0.7).add_to(map_final)
       
map_final